### PR TITLE
Add CTE Materialization documentation

### LINKED ISSUE #22648: Add CTE materialization User documentation

### PR DESCRIPTION
**Pull request by jaystarshot**
## Description
Fixes #22648

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* ... :pr:`12345`
* ... :pr:`12345`

Hive Connector Changes
* ... :pr:`12345`
* ... :pr:`12345`
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by jaystarshot** [comment]
cc: @feilong-liu @mlyublena @tdcmeehan @rschlussel  for any suggestions

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 1]
```diff
@@ -0,0 +1,103 @@
+=================================
+CTE Materialization
+=================================
+
+CTEs (Common Table Expressions) are defined as subqueries that appear in a WITH clause provided by the user.
```
Comment (elharo) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:5]: Common Table Expressions (CTEs)

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+Their repeated usage within a query can lead in unnecessary redundant computations, excessive data retrieval, and high resource consumption
```
Comment (elharo) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:6]: lead to
period after consumption

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+To address this, Presto supports CTE Materialization. This feature allows intermediate CTEs to be materialized and effectively stored
```
Comment (elharo) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:8]: "materialized and effectively stored
and reused within" --> materialized, stored, and reused within

or maybe just "reused"

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+and reused within the scope of the same query.
+
+This improves performance and is especially beneficial when the same CTE is utilized multiple times in a query.
+These CTEs will be stored in temporary tables that are bucketed based on the first projection column
```
Comment (elharo) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:12]: period

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+To use this feature the connector used by the query must support the creation of temporary tables. Currently, only the :doc:`/connector/hive` connector offers this capability.
```
Comment (elharo) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:14]: comma after feature

Comment (elharo) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:14]: connector offers --> Hive Connector offers

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+How to use CTE Materialization
+--------------
+
+The following configurations and session properties are provided to enable CTE materialization and fine-tune its settings.
```
Comment (elharo) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:19]: delete "are provided to"

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+``cte-materialization-strategy``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``string``
+    * **Allowed values:** ``ALL``, ``NONE``, ``HEURISTIC`, ``HEURISTIC_COMPLEX_QUERIES_ONLY``
+    * **Default value:** ``NONE``
+
+    Specifies the strategy to use for materializing Common Table Expressions (CTEs) in queries.
```
Comment (elharo) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:28]: delete "to use"

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    ``NONE`` indicates that no CTEs will be materialized.
```
Comment (elharo) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:29]: delete "indicates that"

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    ``ALL`` indicates that all CTEs in the query will be materialized.
```
Comment (elharo) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:30]: delete "indicates that"

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    This can also be specified on a per-query basis using the ``cte_materialization_strategy`` session property.
+
+``cte-heuristic-replication-threshold``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``integer``
+    * **Minimum value:** ``0``
+    * **Default value:** ``4``
+
+    Used with CTE Materialization Strategy is set to ``HEURISTIC` or ``HEURISTIC_COMPLEX_QUERIES_ONLY``. CTES are only materialized if they are used greater than or equal to this number
```
Comment (elharo) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:42]: I don't understand this sentence. Please rewrite

**Review by steveburnett — CHANGES_REQUESTED** [review]
Nice work, thanks for the doc! I made a lot of suggestions that are mostly repetitive. I had a few comments that I ask for your thoughts. 

Also, I'm planning to reformat the Properties Reference page to remove the indents for content under each header. 
Instead of cluttering this review with even more comments, I'll ask you here to remove the indents for the content of the property entries under the headings. 

Here's a screenshot of my local build where I removed the indents for ``cte-materialization-strategy``, showing the format change in the build - removing the gray vertical line - from that entry, but leaving the entry for ``cte-heuristic-replication-threshold`` to show what I'm asking you to do. Let me know if you have any questions. 
 
Screenshot 2024-05-07 at 1 16 26 PM

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``string``
+    * **Allowed values:** ``ALL``, ``NONE``, ``HEURISTIC`, ``HEURISTIC_COMPLEX_QUERIES_ONLY``
+    * **Default value:** ``NONE``
+
+    Specifies the strategy to use for materializing Common Table Expressions (CTEs) in queries.
+    ``NONE`` indicates that no CTEs will be materialized.
+    ``ALL`` indicates that all CTEs in the query will be materialized.
+    ``HEURISTIC` greedily materializes the earliest parent CTE which is repeated >= cte_heuristic_replication_threshold times
```
Comment (aditi-pandit) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:31]: @jaystarshot : Do you intend to have a blog article or some doc giving more examples (queries and plans) demonstrating the behavior of these config ?

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+``query.cte-hash-partition-count``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``integer``
+    * **Default value:** ``100``
+
+    The number of partitions to be used for materializing Common Table Expressions (CTEs) in queries.
+    This setting determines how many buckets or writers should be used when materializing the CTEs, potentially affecting the performance of queries involving CTE materialization.
+    A higher number of partitions might improve parallelism but also increases overhead in terms of memory and network communication.
```
Comment (aditi-pandit) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:53]: Its strange that a property called partitioning actually affects the buckets. But please emphasize in the explanation that these are buckets and not partitions.

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``string``
+    * **Default value:** ``system``
+
+    The name of the catalog to be used and which provides custom partitioning for Common Table Expression (CTE) materialization.
+    This setting specifies which catalog should be used for CTE materialization and for determining how to partition the materialization of CTEs in queries.
+    This can also be specified on a per-query basis using the ``cte_partitioning_provider_catalog`` session property.
+
+``cte-filter-and-projection-pushdown-enabled``
```
Comment (aditi-pandit) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:67]: Would be great to have an example of this as well. Do you have something in a design doc ?

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    This can also be specified on a per-query basis using the ``hive.temporary_table_storage_format`` session property.
+
+``hive.bucket-function-type-for-cte-materialization``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``string``
+    * **Allowed values:** ``HIVE_COMPATIBLE``, ``PRESTO_NATIVE``
+    * **Default value:** ``PRESTO_NATIVE``
+
+    This setting specifies the Hash function type for cte materialization
```
Comment (aditi-pandit) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:94]: What do each of these settings mean ? Can you elaborate ?

**Review by aditi-pandit — COMMENTED** [review]
Thanks @jaystarshot for this doc. Have some comments.

**Review by feilong-liu — DISMISSED** [review]

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (jaystarshot) [reply presto-docs/src/main/sphinx/admin/cte-materialization.rst:31]: Yes its related to greedy vs dynamic programming choice

Comment (jaystarshot) [reply presto-docs/src/main/sphinx/admin/cte-materialization.rst:31]: Yes that is a good idea, i can add it as a blog or https://github.com/prestodb/presto/wiki

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst (hunk 12 above)
Comment (jaystarshot) [reply presto-docs/src/main/sphinx/admin/cte-materialization.rst:53]: This actually comes from exchange materialization where we use the hash partition count to count the number of buckets

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst (hunk 13 above)
Comment (jaystarshot) [reply presto-docs/src/main/sphinx/admin/cte-materialization.rst:67]: https://github.com/prestodb/presto/issues/21638 There is a good example on this ticket

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+    Flag to enable or disable the pushdown of common filters and projects into the materialized CTE.
+    This can also be specified on a per-query basis using the ``cte_filter_and_projection_pushdown_enabled`` session property.
+
+``hive.temporary-table-storage-format``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``string``
+    * **Allowed values:** ``PAGEFILE``, ``ORC``, ``DWRF``, ``ALPHA``, ``PARQUET``, ``AVRO``, ``RCBINARY``, ``RCTEXT``, ``SEQUENCEFILE``, ``JSON``, ``TEXTFILE``, ``CSV``
+    * **Default value:** ``ORC``
```
[the remark this answers is not in the document]
Comment (jaystarshot) [reply presto-docs/src/main/sphinx/admin/cte-materialization.rst:81]: Yeah we can I will do it in a follow up PR. This is also used by exchange materialization and I did not want to break any default flow incase anyone was using the default value. Also prestissimo doesn't support PAGEFILE for now

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+
+    This setting determines the data format for temporary tables generated by CTE materialization. The recommended value is ``PAGEFILE`` :doc:`/develop/serialized-page`, as it is the most performant,
```
[the remark this answers is not in the document]
Comment (jaystarshot) [reply presto-docs/src/main/sphinx/admin/cte-materialization.rst:83]: Same as above^ This is also used by exchange materialization and I did not want to break any default flow incase anyone was using the default value. Also prestissimo doesn't support PAGEFILE for now

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst (hunk 14 above)
Comment (jaystarshot) [reply presto-docs/src/main/sphinx/admin/cte-materialization.rst:94]: https://github.com/prestodb/presto/commit/cb69e899efbabf35a09003852194800309f53dad
Sample commit which added them. 


Its used https://github.com/prestodb/presto/blob/785ece64987d21d465ee66731e0d39282754e9bb/presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java#L2968
I think it determines which hashing functions to use when inserting records into the buckets.

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+
+    * **Type:** ``string``
+    * **Allowed values:** ``ALL``, ``NONE``, ``HEURISTIC``, ``HEURISTIC_COMPLEX_QUERIES_ONLY``
+    * **Default value:** ``NONE``
+
+    Specifies the strategy for materializing Common Table Expressions (CTEs) in queries.
+    ``NONE`` - no CTEs will be materialized.
+    ``ALL``  - all CTEs in the query will be materialized.
+    ``HEURISTIC`` - greedily materializes the earliest parent CTE, which is repeated >= ``cte_heuristic_replication_threshold`` times.
+    ``HEURISTIC_COMPLEX_QUERIES_ONLY` greedily materializes the earliest parent CTE which meets the ``HEURISTIC`` criteria and has a join or aggregate
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:30]:
```suggestion
    ``HEURISTIC_COMPLEX_QUERIES_ONLY`` greedily materializes the earliest parent CTE which meets the ``HEURISTIC`` criteria and has a join or aggregate
```

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+    Use the ``cte_materialization_strategy`` session property to set on a per-query basis.
+
+``cte-heuristic-replication-threshold``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``integer``
+    * **Minimum value:** ``0``
+    * **Default value:** ``4``
+
+    When ``cte-materialization-strategy`` is set to ``HEURISTIC`` or ``HEURISTIC_COMPLEX_QUERIES_ONLY``, then CTEs will be materialized if they are appear in a query at least``cte-heuristic-replication-threshold`` number of times.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:40]:
```suggestion
    When ``cte-materialization-strategy`` is set to ``HEURISTIC`` or ``HEURISTIC_COMPLEX_QUERIES_ONLY``, then CTEs will be materialized if they appear in a query at least ``cte-heuristic-replication-threshold`` number of times.
```

**Review by steveburnett — CHANGES_REQUESTED** [review]
Nice! A couple of nits for formatting issues, but from a docs standpoint it looks good. Happy to take another look when you're done.

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+``cte-materialization-strategy``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``string``
+    * **Allowed values:** ``ALL``, ``NONE``, ``HEURISTIC``, ``HEURISTIC_COMPLEX_QUERIES_ONLY``
+    * **Default value:** ``NONE``
+
+    Specifies the strategy for materializing Common Table Expressions (CTEs) in queries.
+    ``NONE`` - no CTEs will be materialized.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:27]: These entries - ``NONE``, ``ALL``, ``HEURISTIC``, and ``HEURISTIC_COMPLEX_QUERIES_ONLY`` - look neat in the code here, but when compiled they all pile together in a paragraph. Add a line break before each item to left-justify them.

Screenshot to show the current state. 
Screenshot 2024-05-13 at 2 32 32 PM copy

**Review by steveburnett — CHANGES_REQUESTED** [review]
Thanks for the quick fix! In addition to the one comment about the value descriptions in cte-materialization-strategy, I have a request that would be a repetitive set of changes and is simpler to make here. 

Suggest removing the indent throughout to remove the vertical stripe on the left. I'm trying to get to this throughout the Presto doc and it's taking time, but your new doc following this behavior would help. 

Screenshots of the first one and a half entries (before, and after - the after also includes my suggested formatting for the first item's values) from my local environment to show what I mean. 

(The second entry is unchanged in after, only cte-materialization-strategy acts as an example.)

before
Screenshot 2024-05-13 at 2 32 32 PM


after
Screenshot 2024-05-13 at 2 34 17 PM

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+CTE Materialization
+===================
+
+Common Table Expressions (CTEs) are subqueries that appear in a WITH clause provided by the user.
+Their repeated usage in a query can lead to redundant computations, excessive data retrieval, and high resource consumption.
+
+To address this, Presto supports CTE Materialization allowing intermediate CTEs to be reused within the scope of the same query.
+Materializing CTEs can improve performance when the same CTE is used multiple times in a query by reducing recomputation of the CTE. However, there is also a cost to writing to and reading from disk, so the optimization may not be beneficial for very simple CTEs
+or CTEs that are not used many times in a query.
+Materialized CTEs are stored in temporary tables that are bucketed based on the first projection column.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:11]:
```suggestion

Materialized CTEs are stored in temporary tables that are bucketed based on the first projection column.
```
Suggest a line break to left-justify this sentence and break up this paragraph to improve readability.

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+
+Specifies the strategy for materializing Common Table Expressions (CTEs) in queries.
+
+``NONE`` - no CTEs will be materialized.
+
+``ALL``  - all CTEs in the query will be materialized.
+
+``HEURISTIC`` - greedily materializes the earliest parent CTE, which is repeated >= ``cte_heuristic_replication_threshold`` times.
+
+``HEURISTIC_COMPLEX_QUERIES_ONLY`` greedily materializes the earliest parent CTE which meets the ``HEURISTIC`` criteria and has a join or aggregate
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:35]:
```suggestion
``HEURISTIC_COMPLEX_QUERIES_ONLY`` greedily materializes the earliest parent CTE which meets the ``HEURISTIC`` criteria and has a join or aggregate.
```

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+
+    * **Type:** ``integer``
+    * **Default value:** ``100``
+
+The number of buckets to be used for materializing CTEs in queries.
+This setting determines how many buckets or writers should be used when materializing the CTEs, potentially affecting the performance of queries involving CTE materialization.
+A higher number of partitions might improve parallelism but also increases overhead in terms of memory and network communication.
+
+Recommended value: 4 - 10x times the size of the cluster.
+Use the ``cte_hash_partition_count`` session property to set on a per-query basis.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:60]:
```suggestion

Use the ``cte_hash_partition_count`` session property to set on a per-query basis.
```

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+
+``query.cte-partitioning-provider-catalog``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``string``
+    * **Default value:** ``system``
+
+The name of the catalog that provides custom partitioning for CTE materialization.
+This setting specifies which catalog should be used for CTE materialization.
+Use the ``cte_partitioning_provider_catalog`` session property to set on a per-query basis.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:70]:
```suggestion

Use the ``cte_partitioning_provider_catalog`` session property to set on a per-query basis.
```

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+
+``cte-filter-and-projection-pushdown-enabled``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``boolean``
+    * **Default value:** ``true``
+
+Flag to enable or disable the pushdown of common filters and projects into the materialized CTE.
+Use the ``cte_filter_and_projection_pushdown_enabled`` session property to set on a per-query basis.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:79]:
```suggestion

Use the ``cte_filter_and_projection_pushdown_enabled`` session property to set on a per-query basis.
```

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+``hive.temporary-table-storage-format``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``string``
+    * **Allowed values:** ``PAGEFILE``, ``ORC``, ``DWRF``, ``ALPHA``, ``PARQUET``, ``AVRO``, ``RCBINARY``, ``RCTEXT``, ``SEQUENCEFILE``, ``JSON``, ``TEXTFILE``, ``CSV``
+    * **Default value:** ``ORC``
+
+This setting determines the data format for temporary tables generated by CTE materialization. The recommended value is ``PAGEFILE`` :doc:`/develop/serialized-page`, as it is the most performant,
+since it avoids serialization and deserialization during reads and writes, allowing for direct storage of Presto pages.
+Use the ``hive.temporary_table_storage_format`` session property to set on a per-query basis.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:90]:
```suggestion

Use the ``hive.temporary_table_storage_format`` session property to set on a per-query basis.
```

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+
+``hive.bucket-function-type-for-cte-materialization``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``string``
+    * **Allowed values:** ``HIVE_COMPATIBLE``, ``PRESTO_NATIVE``
+    * **Default value:** ``PRESTO_NATIVE``
+
+This setting specifies the Hash function type for CTE materialization.
+Use the ``hive.bucket_function_type_for_cte_materialization`` session property to set on a per-query basis.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:100]:
```suggestion

Use the ``hive.bucket_function_type_for_cte_materialization`` session property to set on a per-query basis.
```

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+
+
+``query.max-written-intermediate-bytes``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``DataSize``
+    * **Default value:** ``2TB``
+
+This setting defines a cap on the amount of data that can be written during CTE Materialization. If a query exceeds this limit, it will fail.
+Use the ``query_max_written_intermediate_bytes`` session property to set on a per-query basis.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:110]:
```suggestion

Use the ``query_max_written_intermediate_bytes`` session property to set on a per-query basis.
```

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst (hunk 12 above)
Comment (steveburnett) [reply presto-docs/src/main/sphinx/admin/cte-materialization.rst]: Thanks for explaining! I thought this was strange, but the explanation makes sense.

File: presto-docs/src/main/sphinx/admin/cte-materialization.rst [hunk 28]
```diff
@@ -0,0 +1,118 @@
...
+
+    * **Type:** ``string``
+    * **Default value:** ``system``
+
+The name of the catalog that provides custom partitioning for CTE materialization.
+This setting specifies which catalog should be used for CTE materialization.
+Use the ``cte_partitioning_provider_catalog`` session property to set on a per-query basis.
+
+``cte-filter-and-projection-pushdown-enabled``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/cte-materialization.rst:73]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Review by steveburnett — CHANGES_REQUESTED** [review]
Nice work! Only a few nits.

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)

Pulled updated branch, new local build of docs to review, everything looks good. Thanks!

**Pull request merged**